### DISCUSSION TITLE
Solving for eigenvectors; definition of memory kernel

### DISCUSSION THREAD

**Discussion by thvoigtmann** [IlianPihlajamaa/ModeCouplingTheory.jl#35] [General]
So, I've started to implement code to solve for the critical eigenvectors of MCT, from which we can finally get the exponent parameter.

Have a look at this changeset:
https://github.com/IlianPihlajamaa/ModeCouplingTheory.jl/commit/1e8d8b9da5009d5f1681cccde7ff13e1bf921338

(I created a branch 'eigenvectors' in my fork, but I forgot to switch back to main when checking out the new branch, so now it's based on the beta-scaling. If you're proficient enough in git to tell me how to change that ('eigenvectors' should branch off 'main'), I'll be happy.)

I wanted to discuss with you two aspects of the code structure:

- (1) Implementation of the eigenvalue solver: I extended the code for `ModeCouplingKernel` to allow for `evaluate_kernel_linearization`. To remind you what we need to do: if the memory kernel is a nonlinear functional, we need the linearization $C = \partial K[F]/\partial F$, called the stability matrix, evaluated at the critical point. For a bilinear memory kernel, we have $C[\cdot]=K[F^c,\cdot]$ up to prefactors (some of which are conventional). This means we can reuse the code for the memory kernel, in particular `fill_A!` for the Bengtzelius trick, if we allow for the two arguments to be different. I am hoping that my change to `fill_A!` doesn't induce a performance penalty. What do you think?
- I thought about creating a new struct to represent linearizations of kernels, something like `Linearization{ModeCouplingKernel} <: ModeCouplingKernel` which would allow to simply declare new methods for the linearization, such that the linearization object could also store its reference point $F^c$. But I'm not sure that is how julia works.
- I added code for the right eigenvector: this is solved by iterating $H^{(n+1)}=K[F^c,H^{(n)}]$ (again omitting prefactors) and normalizing after each step by $\|H^{(n+1)}\|$. That norm will actually be our estimate of the eigenvalue. We also need the left eigenvector, that is the vector $\hat H$ that fulfills $(\hat H,K[F^c,G])=(\hat H,G)$ (up to prefactors) for each $G$. This means, we need to calculate not $\sum_{qp}$ a la Bengtzelius, but $\sum_{kq}$. I'll look into that once I've understood your version of the Bengtzelius trick, which seems to be different from mine (and simpler).
- (2) More crucially: I would really like to discuss with you the definition of $K$, regarding whether we should include $\gamma$ or not. You remember our discussion in Eindhoven: in the equations that determine $F^c$ and all the properties of the transition, the masses should drop out exactly. So it would be good and also semantically cleaner if the user could solve for the "steady state" properties without having to specify $\gamma$ (or rather, $k_BT/m$). I wish to propose to change the definition of "memory kernel" throughout the code, but I'm happy to hear arguments against that.
- I think that `solve_steady_state` is slightly misleadingly named. As long as we calculate correlation functions that are time-translation invariant, we are always in steady state. I propose `solve_nonergodicity_parameters` or `solve_longtime_limit` or something like that.

I'll also be happy if you can review my code. I'm not entirely sure I'm handling the distinction between `Vector` and `Diagonal` in the most efficient way, because I haven't yet understood completely at which points along the chain of function calls you promote and demote vectors to and from diagonal matrices.

**Comment by IlianPihlajamaa** [comment]
There is a lot to unpack here, so let me do it one at a time:

**Comment by IlianPihlajamaa** [comment]
> I created a branch 'eigenvectors' in my fork, but I forgot to switch back to main when checking out the new branch, so now it's based on the beta-scaling. If you're proficient enough in git to tell me how to change that ('eigenvectors' should branch off 'main'), I'll be happy.

I am by no means a git wizard, but it sounds like you need `git rebase`. See here: https://stackoverflow.com/questions/3810348/setting-git-parent-pointer-to-a-different-parent 
I can also just merge your beta-scaling work after we settle the `NoMemTempStruct` thing, and then you can open a new PR for this.

**Comment by IlianPihlajamaa** [comment]
> (1) Implementation of the eigenvalue solver: I extended the code for ModeCouplingKernel to allow for evaluate_kernel_linearization. To remind you what we need to do: if the memory kernel is a nonlinear functional, we need the linearization, called the stability matrix, evaluated at the critical point. For a bilinear memory kernel, we have 
>  up to prefactors (some of which are conventional). This means we can reuse the code for the memory kernel, in particular fill_A! for the Bengtzelius trick, if we allow for the two arguments to be different. I am hoping that my change to fill_A! doesn't induce a performance penalty. What do you think?

Sounds good! I don't think that would cause any performance penalty. To be sure, I will check tomorrow.

**Comment by IlianPihlajamaa** [comment]
> I thought about creating a new struct to represent linearizations of kernels, something like Linearization{ModeCouplingKernel} <: ModeCouplingKernel which would allow to simply declare new methods for the linearization, such that the linearization object could also store its reference point $F_c$. But I'm not sure that is how julia works.

I think we have a choice of API here. 
1. Either we make a new struct `LinearizedKernel{ModeCouplingKernel}` as you suggest, and call `evaluate_kernel!(out, linearized_kernel::LinearizedKernel, Fc, H)` to evaluate it (perhaps without the `Fc` if it is contained within the `linearized_kernel`)
2. We add a new function `evaluate_linearized_kernel!(out, kernel::MemoryKernel, Fc, H)` that, given the type of the `kernel`, does the right thing. We then just need to write the right methods for it.

What do you think?

**Comment by IlianPihlajamaa** [comment]
>  I added code for the right eigenvector: this is solved by iterating $H^{(n+1)}=K[F^c,H^{(n)}]$ (again omitting prefactors) and normalizing after each step by $\|H^{(n+1)}\|$. That norm will actually be our estimate of the eigenvalue. We also need the left eigenvector, that is the vector $\hat H$ that fulfils $(\hat H,K[F^c,G])=(\hat H,G)$ (up to prefactors) for each $G$. This means, we need to calculate not $\sum_{qp}$ a la Bengtzelius, but $\sum_{kq}$. I'll look into that once I've understood your version of the Bengtzelius trick, which seems to be different from mine (and simpler).

Great, thank you! I will have a more detailed look tomorrow. For now, here are my personal notes on it, maybe that explains my method. From what I could tell from yours (I only quickly skimmed over that part), what we do is essentially the same. 

[Discretisation_of_memory_kernel.pdf](https://github.com/IlianPihlajamaa/ModeCouplingTheory.jl/files/10475033/Discretisation_of_memory_kernel.pdf)

**Comment by IlianPihlajamaa** [comment]
> (2) More crucially: I would really like to discuss with you the definition of $K$, regarding whether we should include $\gamma$ or not. You remember our discussion in Eindhoven: in the equations that determine $F^c$ and all the properties of the transition, the masses should drop out exactly. So it would be good and also semantically cleaner if the user could solve for the "steady state" properties without having to specify $\gamma$ (or rather, $k_BT/m$). I wish to propose to change the definition of "memory kernel" throughout the code, but I'm happy to hear arguments against that.

I remember our conversation, and I get your point, but I am unsure how to address this generally. 
Are you suggesting that, instead of solving 
$$a\ddot{F}+b\dot{F}+cF+d+\int_\tau K(t-\tau)\dot{F}(\tau)d\tau=0$$
we instead only solve for
$$c^{-1}a\ddot{F}+c^{-1}b\dot{F}+F+c^{-1}d+c^{-1}\int_\tau K(t-\tau)\dot{F}(\tau)d\tau=0,$$
so that the $k_BT/m$ drops out when we solve for the long time limit of MCT specifically?

Or do you suggest that we define the MCT memory kernel by what is now c^{-1}K(t-\tau), so that the frequency coefficient becomes the identity matrix? 

In any case, I would like our code to work with a generic GLE,
$$a\ddot{F}+b\dot{F}+cF+d+\int_\tau K(t-\tau)\dot{F}(\tau)d\tau=0$$
without restricting $c=I$. And as I see it, the long time limit, in the general case, depends on that coefficient, so `find_nonergodicity_parameter` has to take $c$ as an input. I don't really see a way around that.

**Comment by IlianPihlajamaa** [comment]
> I think that solve_steady_state is slightly misleadingly named. As long as we calculate correlation functions that are time-translation invariant, we are always in steady state. I propose solve_nonergodicity_parameters or solve_longtime_limit or something like that.

The reason I chose not to go for `find/solve_non_ergodicity_parameter(s)` is that I did not want to impose too much physical interpretation on the user. I know that this may never happen, but there may be scenarios in which the plateau value really has nothing to do with ergodicity-breaking, leading to confusion. Though I know that steady-state is ambiguous physically, in the mathematical sense, it means exactly what we use it for.

I think `solve_long_time_limit` is a good middle ground. I would agree to change to that.
